src/index.ts

```diff
@@ -1 +1,73 @@
-//Define class here
\ No newline at end of file
+export interface IUser {
+  name: string
+  favoriteColor: string
+  age: number
+  readonly id?: string
+}
+
+export class UserAPI {
+
+  constructor(private _users: Record<string, IUser> = {}) {
+
+  }
+
+  addUser(user: IUser): IUser {
+    const id: Readonly<string> = Date.now().toString()
+    if (user.hasOwnProperty("id")) {
+      throw new Error("Id incorrectly provided by input user")
+    }
+    if (Object.values(this._users).some(
+      x => x.name.toLowerCase() === user.name.toLowerCase() &&
+        x.favoriteColor.toLowerCase() === x.favoriteColor.toLowerCase() &&
+        x.age === user.age
+    )) {
+      throw new Error("User with these properties already exists")
+    }
+    this._users = { ...this._users, [id]: { ...user, id: id } }
+    return { ...user, id: id }
+  }
+
+  getUserById(id: string): Readonly<IUser> {
```

Comment:
```suggestion
  getUserById(id: string): IUser {
```

because `IUser` is readonly (or will be once all of its properties are `readonly`), we don't need `Readonly` here.

there is a pattern you might see in some places where a declared interface extends an empty Readonly object, so that the properties don't have to be individually declared as `readonly`.  That pattern would also be acceptable